middlewares/authzCourseOrInstance.js

```diff
@@ -0,0 +1,485 @@
+                                        `All requested changes to the effective user have been removed.</p>`;
+                            return callback(err);
+                        }
+                    }
+
+                    res.locals.authz_data.user = user;
+                    res.locals.authz_data.is_administrator = is_administrator;
+                    res.locals.authz_data.course_role = result.rows[0].permissions_course.course_role;
+                    res.locals.authz_data.has_course_permission_preview = result.rows[0].permissions_course.has_course_permission_preview;
+                    res.locals.authz_data.has_course_permission_view = result.rows[0].permissions_course.has_course_permission_view;
+                    res.locals.authz_data.has_course_permission_edit = result.rows[0].permissions_course.has_course_permission_edit;
+                    res.locals.authz_data.has_course_permission_own = result.rows[0].permissions_course.has_course_permission_own;
+
+                    // Effective users are confined to one course, so we discard all other
+                    // courses from the list of those to which the effective user has staff
+                    // access.
+                    //
+                    // Note that courses[0].permissions_course and course.permissions_course
+                    // will not, in general, be the same. Requested course and course instance
+                    // roles are ignored when generating the former but not when generating
+                    // the latter. So, we also replace permissions_course for the course that
+                    // remains in the list (the current course) with what it should be.
+                    //
+                    // We then update editable_courses as usual.
+                    res.locals.authz_data.courses = (result.rows[0].courses || []).filter(course => course.id == result.rows[0].course.id);
+                    res.locals.authz_data.courses.forEach(course => course.permissions_course = _.cloneDeep(result.rows[0].permissions_course));
+                    res.locals.authz_data.editable_courses = res.locals.authz_data.courses.filter(course => course.permissions_course.has_course_permission_edit);
+
+                    // Update course_instances, adding a flag to disable any course
+                    // instance that is not also in authn_course_instances (i.e., to
+                    // which the authn user does not also have access)
+                    res.locals.authz_data.course_instances = result.rows[0].course_instances || [];
+                    res.locals.authz_data.course_instances.forEach((ci) => {
+                        ci.disabled = (! res.locals.authz_data.authn_course_instances.some((authn_ci) => {
+                            return ci.id == authn_ci.id;
+                        }));
+                    });
```

Comment:
The only place this is used is here (and on the following line): https://github.com/PrairieLearn/PrairieLearn/blob/65893e2c139d966757d2f126ca6811904725a6b7/pages/partials/navbarInstructor.ejs#L45

Reply:
I think that keeping the non-authn-CIs (even if disabled) is an information leak, because it allows an authn user to see which CIs another user has access to.

Reply:
I changed this in 98b311f7cf9dd33f8708f6dc9be2dc0870e33fc5 so that we now set the list of course instances to be the _intersection_ of the course instances from the effective and authn users.